crates/sui-core/src/authority/authority_notify_read.rs

```diff
@@ -0,0 +1,209 @@
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use sui_types::base_types::TransactionDigest;
+use sui_types::error::SuiResult;
+use sui_types::messages::TransactionEffects;
+use tokio::sync::oneshot;
+
+#[allow(dead_code)]
+#[async_trait]
+pub trait EffectsNotifyRead {
+    /// This method reads transaction effects from database.
+    /// If effects are not available immediately, the method blocks until they are persisted
+    /// in the database.
+    ///
+    /// This method **does not** schedule transactions for execution - it is responsibility of the caller
+    /// to schedule transactions for execution before calling this method.
+    async fn notify_read(
+        &self,
+        digests: Vec<TransactionDigest>,
+    ) -> SuiResult<Vec<TransactionEffects>>;
+}
+
+type Registrations<V> = Vec<oneshot::Sender<V>>;
+
+pub(crate) struct NotifyRead<K, V> {
+    pending: Vec<Mutex<HashMap<K, Registrations<V>>>>,
+    count_pending: AtomicUsize,
+}
+
+impl<K: Eq + Hash + Clone, V: Clone> NotifyRead<K, V> {
+    pub fn new() -> Self {
+        let pending = (0..255).map(|_| Default::default()).collect();
+        let count_pending = Default::default();
+        Self {
+            pending,
+            count_pending,
+        }
+    }
+
+    /// Asynchronously notifies waiters and return number of remaining pending registration
+    pub fn notify(&self, key: &K, value: &V) -> usize {
+        let registrations = self.pending(key).remove(key);
+        let Some(registrations) = registrations else {
+            return self.count_pending.load(Ordering::Relaxed);
+        };
+        let rem = self
+            .count_pending
+            .fetch_sub(registrations.len(), Ordering::Relaxed);
+        for registration in registrations {
+            registration.send(value.clone()).ok();
+        }
+        rem
+    }
+
+    pub fn register_all(&self, keys: Vec<K>) -> Vec<Registration<K, V>> {
+        self.count_pending.fetch_add(keys.len(), Ordering::Relaxed);
+        let mut registrations = vec![];
+        for key in keys.iter() {
+            let (sender, receiver) = oneshot::channel();
+            self.register(key, sender);
+            let registration = Registration {
+                this: self,
+                registration: Some((key.clone(), receiver)),
+            };
+            registrations.push(registration);
+        }
+        registrations
+    }
+
+    fn register(&self, key: &K, sender: oneshot::Sender<V>) {
+        self.pending(key)
+            .entry(key.clone())
+            .or_default()
+            .push(sender);
+    }
+
+    fn pending(&self, key: &K) -> MutexGuard<HashMap<K, Registrations<V>>> {
+        let mut state = DefaultHasher::new();
+        key.hash(&mut state);
+        let hash = state.finish();
+        let pending = self
+            .pending
+            .get((hash % self.pending.len() as u64) as usize)
+            .unwrap();
+        pending.lock()
```

Comment:
lock returns a LockResult... I don't understand how this is compiling.

Reply:
I am using parking_lot locks, they don't return error on lock (only std:: locks do that)